### ISSUE TITLE
`RequestGuildPartyState()` can possibly contribute to, or help with determining the cause of the recent disconnects

### ISSUE THREAD

**Issue by Ghostopheles** [Stanzilla/WoWUIBugs#442]
tldr; this function and the circumstances it exists in are prone to weird behavior and bugs and can be used as markers to possibly help with disconnects. The disconnect specifically DCs you from the Home server, which is where guild and chat info is handled. If nothing else, the below info will be useful for reproducing the issue and further testing.

If you /reload quickly (but not *too* quickly) after loading into the game for the first time (not a /reload or relog, has to be a cold start), the `MinimapCluster.InstanceDifficulty` frame will (slightly inconsistently) get stuck in a broken state, and after a few seconds you will be disconnected from the game. This error seems to repeat itself semi-consistently after it's occurred once. Restarting the game breaks the cycle most of the time, though you may get unlucky and encounter it over and over again like I do. 

Also, before I go too far down the rabbit hole, I'd like to acknowledge that this is far from concrete evidence and it could be that this function is innocent and is just falling victim to a completely separate issue that's causing players to disconnect. Whatever the cause may be, hopefully this can point someone in the direction of a fix.

[InstanceDifficulty.lua](https://github.com/Gethe/wow-ui-source/blob/live/Interface/FrameXML/InstanceDifficulty.lua)

If you mouse over the `InstanceDifficulty` frame while it's in the broken state, triggering it's `OnEnter` script, it'll trigger bug #405.

This bug appears to be impacting a very large number of players and I would consider it fairly critical.

# Details
Upon loading, the frame registers a couple of events. Most of which occur before my addon can load (or just otherwise do not want to be observed), but I have a suspicion that it's either [PARTY_MEMBER_ENABLE](https://wowpedia.fandom.com/wiki/PARTY_MEMBER_ENABLE) or [PARTY_MEMBER_DISABLE](https://wowpedia.fandom.com/wiki/PARTY_MEMBER_DISABLE), both of which will trigger the `RequestGuildPartyState()` function call [here.](https://github.com/Gethe/wow-ui-source/blob/e8693cc8bc1f4ca2391216badfcf63290328aefd/Interface/FrameXML/InstanceDifficulty.lua#L40)

It's also possible that it's being triggered by [UPDATE_INSTANCE_INFO](https://wowpedia.fandom.com/wiki/UPDATE_INSTANCE_INFO), [INSTANCE_GROUP_SIZE_CHANGED](https://wowpedia.fandom.com/wiki/INSTANCE_GROUP_SIZE_CHANGED), or [PLAYER_GUILD_UPDATE](https://wowpedia.fandom.com/wiki/PLAYER_GUILD_UPDATE), though it doesn't really matter what's triggering it, as none of the functions that are supposed to follow ever  occur. This leads me to believe that `RequestGuildPartyState()` is failing internally and for whatever reason doesn't allow the subsequent code to run, and at the same time causes the server to yeet your client into the shadow realm.

The disconnect is also not your average disconnect. You lose connection specifically to the Home/Realm server, which handles guild info, chat, and a few other things, so while you're in this state you can still see players running around.
![image](https://github.com/Stanzilla/WoWUIBugs/assets/10636803/31277cfa-910b-4154-a7a7-0d9bae6514bc)
After a few seconds of not being connected, you get kicked out. If you log out while you're in this half-disconnected state, you'll be taken to the character select screen but nothing will load, and eventually you will disconnect.

I'm fairly strongly convinced that `RequestGuildPartyState()` is the culprit, though it's possible I'm wrong. Regardless though, when the bug occurs it isn't triggering my posthook, so either it gets stuck or some other catastrophic event occurs.

One more note, the function `RequestGuildPartyState()` is only used in [InstanceDifficulty.lua](https://github.com/Gethe/wow-ui-source/blob/live/Interface/FrameXML/InstanceDifficulty.lua) and is not referenced anywhere else in the UI.

# Examples
My event listener was listening to all of the same events the `InstanceDifficulty` frame was, and I had hooked the functions in question to print out when they were called. All the below screenshots were taken **after** doing the quick-reload-after-loading-in thing.

![image](https://github.com/Stanzilla/WoWUIBugs/assets/10636803/7060a1e1-85dd-4afe-9a8a-87e1d47fce9e)
This is the baseline I used as what a nominal sequence should look like. The first two `GetInstanceInfo()` calls are made by UnitFrames and can be ignored. The rest of the functions from `RequestGuildPartyState()` and on are called from `MinimapCluster.InstanceDifficulty:OnEvent()`.

--

![image](https://github.com/Stanzilla/WoWUIBugs/assets/10636803/8b47e413-8a90-4224-a9ce-5d5584c9dc72)
As you can see here, right after the UnitFrames call `GetInstanceInfo()`, and where `RequestGuildPartyState()` is supposed to run, it instead just stops all execution, then a few moments later, disconnect. And if that wasn't enough pain, the `InstanceDifficulty` banner is also broken! 
![image](https://github.com/Stanzilla/WoWUIBugs/assets/10636803/686521d8-dc4f-46ab-81ae-7c6417ec6e45)

Again, this could just be because the connection was lost prior to the function call and this function just happened to run into it and be visible.
# Tips for reproducing
This one is slightly stubborn but I was able to mostly-somewhat-consistently get it to occur by reloading ~1-2 seconds after initially loading into the world. You'll know the bug occurred if the `InstanceDifficulty` banner on the minimap is bugged like the image above.

Anecdotally, it also seems to help if you transition zones or subzones immediately after loading in, then reload. i.e. moving from Lion's Rest in Stormwind to Stormwind Harbour, then /reload'ing.

This also is not the only way to repro this issue, there are many different ways this issue can occur. Even if this function is not the culprit, the repro methods do still apply.

Blizzard, please fix this urgently. It's one of the biggest complaints I've noticed in the community recently, and even if this isn't 100% accurate, it's a start. :copium:

**Comment by Ghostopheles** [comment]
this was a blunder of epic proportions, might still contain a hint of useful info somewhere though.
